cassiopeia/simulator/SupercellularSampler.py

```diff
@@ -0,0 +1,126 @@
+
+        Args:
+            ratio: The number of times to merge as a ratio of the total number
+                of leaves. A ratio of 0.5 indicates the number of merges will be
+                approximately half the number of leaves.
+            number_of_merges: Explicit number of merges to perform.
+        """
+        if (ratio is None) == (number_of_merges is None):
+            raise LeafSubsamplerError(
+                "Exactly one of 'ratio' and 'number_of_merges' must be specified."
+            )
+
+        self.__ratio = ratio
+        self.__number_of_merges = number_of_merges
+
+    def subsample_leaves(
+        self,
+        tree: CassiopeiaTree,
+        collapse_source: Optional[str] = None,
+        collapse_duplicates: bool = True,
+    ):
+        """Construct a new CassiopeiaTree by merging leaves in the provided
+        tree.
+
+        Args:
+            tree: The CassiopeiaTree for which to subsample leaves
+            collapse_source: The source node from which to collapse unifurcations
+            collapse_duplicates: Whether or not to collapse duplicated character
+                states, so that only unique character states are present in each
+                ambiguous state. Defaults to True.
+
+        """
+        n_merges = (
+            self.__number_of_merges
+            if self.__number_of_merges is not None
+            else int(tree.n_cell * self.__ratio)
+        )
+        if n_merges >= len(tree.leaves):
+            raise LeafSubsamplerError(
+                "Number of required merges exceeds number of leaves in the tree."
+            )
+        if n_merges == 0:
+            raise LeafSubsamplerError("No merges to be performed.")
+        # Tree needs to have character matrix defined
+        tree.get_current_character_matrix()
+
+        merged_tree = copy.deepcopy(tree)
+        for _ in range(n_merges):
+            # Choose first leaf
+            leaf1 = np.random.choice(merged_tree.leaves)
+            leaf1_state = merged_tree.get_character_states(leaf1)
+
+            # Choose second leaf with weight proportional to inverse distance
+            distances = merged_tree.get_distances(leaf1, leaves_only=True)
+            leaves = []
+            weights = []
+            for leaf, distance in distances.items():
+                if leaf == leaf1:
+                    continue
+                leaves.append(leaf)
+                weights.append(1 / distance)
+            leaf2 = np.random.choice(
+                leaves, p=np.array(weights) / np.sum(weights)
+            )
+            leaf2_state = merged_tree.get_character_states(leaf2)
+
+            # Merge these two leaves at the mean time of the two leaves
+            # If the tree is ultrametric, this preserves ultrametricity
+            new_leaf = f"{leaf1}-{leaf2}"
+            new_time = (
+                merged_tree.get_time(leaf1) + merged_tree.get_time(leaf2)
+            ) / 2
```

Comment:
If the tree isn't ultrametric, is there a chance that this new time will be _before_ the ancestor of the two nodes? If so, what's your fix?

Reply:
Good point. I'll choose the average time or the time of the LCA, whichever is greater.

Reply:
Hold up, since we are adding the leaf to the LCA, which MUST have time less than or equal to the two chosen leaves. Therefore, this can never happen.